signer/signer.go

```diff
@@ -0,0 +1,262 @@
+package signer
+
+import (
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	nx509 "github.com/notaryproject/notation-core-go/x509"
+)
+
+type SignatureMediaType string
+
+// SignatureAlgorithm lists supported Signature algorithms.
+type SignatureAlgorithm string
+
+const (
+	RSASSA_PSS_SHA_256 SignatureAlgorithm = "RSASSA_PSS_SHA_256"
+	RSASSA_PSS_SHA_384 SignatureAlgorithm = "RSASSA_PSS_SHA_384"
+	RSASSA_PSS_SHA_512 SignatureAlgorithm = "RSASSA_PSS_SHA_512"
+	ECDSA_SHA_256      SignatureAlgorithm = "ECDSA_SHA_256"
+	ECDSA_SHA_384      SignatureAlgorithm = "ECDSA_SHA_384"
+	ECDSA_SHA_512      SignatureAlgorithm = "ECDSA_SHA_512"
+)
+
+// SignerInfo represents a parsed signature envelope that is agnostic to signature envelope format.
+type SignerInfo struct {
+	Payload            []byte
+	PayloadContentType string
+	SignedAttributes   SignedAttributes
+	UnsignedAttributes UnsignedAttributes
+	SignatureAlgorithm SignatureAlgorithm
+	CertificateChain   []*x509.Certificate
+	Signature          []byte
+	TimestampSignature []byte
+}
+
+// SignedAttributes represents signed metadata in the Signature envelope
+type SignedAttributes struct {
+	SigningTime        time.Time
+	Expiry             time.Time
+	ExtendedAttributes []Attribute
```

Comment:
What if we have two attributes with the same key?

Reply:
If we have two attributes with same key we return two attributes. Its JWS that restricts protected headers to not have duplicate keys but other formats like [PKCS7](https://datatracker.ietf.org/doc/html/rfc2315#section-9.2) allows duplicate keys in attributes.